// backend/schemas/event.schema.cjs
// backend/schemas/event.schema.cjs

// Ajv JSON Schema for POST /api/events
// Matches the payload your frontend sends from BatchDetails.jsx
// {
//   batchId: "BATCH-...",
//   eventType: "Harvested",
//   actor: "farmer",
//   payload: { notes: "..." },
//   createdAt: "2025-11-10T10:00:00.000Z"  // or a number (ms)
// }

const EventSchema = {
  type: "object",
  additionalProperties: false, // only allow the fields we list here
  required: ["batchId", "eventType", "actor"],
  properties: {
    batchId: { type: "string", minLength: 1 },
    certificateID: { type: "string" },        // optional
    eventType: { type: "string", minLength: 1 },
    actor: { type: "string", minLength: 1 },
    payload: {
      type: "object",
      // allow arbitrary keys inside payload (notes/detail/location/etc)
      additionalProperties: true,
      default: {}
    },
    createdAt: {
      anyOf: [
        { type: "string", format: "date-time" }, // ISO string
        { type: "integer", minimum: 0 }          // ms epoch
      ]
    }
  }
};

module.exports = { EventSchema };
